### PR TITLE
Flow control

### PR DESCRIPTION
**Pull request by saratry**
This PR introduces the backpressure functionality, only for retrieving the stream of events for the single aggregate. The implementation takes advantage of the native backpressure feature provided by GRPC. This PR is still in draft, as more work is needed in terms of documentation and testing

### CODE REVIEW

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/CallStreamObserverDelegator.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package io.axoniq.axonserver.localstorage;
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/CallStreamObserverDelegator.java:1]: move to another package, as this is not specific for localstorage.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java [hunk 2]
```diff
@@ -116,6 +109,54 @@ public SegmentBasedEventStore(EventTypeContext eventTypeContext, IndexManager in
...
+                           aggregateReadTimer.record(readTime, TimeUnit.MILLISECONDS);
+                       }
+                   })
+                   .skipUntil(se -> se.getAggregateSequenceNumber() >= firstSequence)
+                   .takeWhile(se -> se.getAggregateSequenceNumber() < lastSequence) //remember: last sequence excluded
+                   .contextWrite(ctx -> ctx.put(subscription_time, new AtomicLong()));
+    }
+
+    private Flux<SerializedEvent> eventsForPositions(long segment, List<Integer> positions) {
+        return (!containsSegment(segment) && next != null) ?
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java:147]: What if !containsSegment and next == null?

File: axonserver/src/main/java/io/axoniq/axonserver/message/event/EventStore.java [hunk 3]
```diff
@@ -57,14 +59,51 @@
...
-    void listAggregateEvents(String context, Authentication authentication, GetAggregateEventsRequest request,
-                             StreamObserver<SerializedEvent> responseStreamObserver);
+    default Flux<SerializedEvent> aggregateSnapshots(String context,
+                                                     Authentication authentication,
+                                                     GetAggregateSnapshotsRequest request) {
+        return Flux.create(
+                sink -> listAggregateSnapshots(context, authentication, request, new StreamObserver<SerializedEvent>() {
+                    @Override
+                    public void onNext(SerializedEvent serializedEvent) {
+                        if (serializedEvent.getAggregateSequenceNumber() < request.getMaxSequence()) {
```
Comment (MGathier) [inline axonserver/src/main/java/io/axoniq/axonserver/message/event/EventStore.java:91]: it should be the responsibility of the listAggregateSnapshots operation to ensure that the sequence number is in the valid range.

**Review by MGathier — COMMENTED** [review]
small comments, but mostly OK

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java (hunk 2 above)
Comment (saratry) [reply axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/SegmentBasedEventStore.java:147]: In this case, we basically return a Flux that completes immediately as the provider of the EventSource returns an empty Optional.

**Pull request merged**